---
title: "0-Prepare_Omics_Data"
author: "Cathy Magnée"
date: "2024-05-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prepare omics data 

This R Markdown document is meant for preparing the metabolomic and transcriptomics (RNA) data used in this project. The data can not be included on the public github, as the article is not published yet. 
This research uses transcriptomic (RNA) and metabolomic data from CD3+ T cells of CLL patients and healthy donors.

Data source: Jacobs, C. F. & Camerini, E. CD3+ T cells from CLL patients and age-matched Healthy Donors. (Future).

The following preparations to the data are in this document:
- Column names are made uniform. 
- Metabolomics
  - HMDB id overview is made

- Transcriptomics
  - RNA FPKM's are imported
  - RNA dataset is split into day 1 and day 2
  - A link is made between ENSEMBL gene ids and gene names

#### Loading packages

Installation dependencies:
```{r}
## Install dependencies
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("ComplexHeatmap")
# 
# ## Stable version, install countToFPKM from CRAN
# install.packages("countToFPKM")
# 
# ## Lastest version, install countToFPKM from GitHub
# if(!require(devtools)) install.packages("devtools")
# devtools::install_github("AAlhendi1707/countToFPKM", build_vignettes = TRUE)
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("biomaRt")

```


```{r, warning=FALSE,message=FALSE}
library(readr)
library(readxl)
library(biomaRt)
library(dplyr)
```

## Loading data

Metabs
```{r, echo=FALSE}
# Path do data: AMC PC
day0_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\DataForStatistics_day0.csv"
day2_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\DataForStatistics_day2.csv"

# Path to data laptop
# day0_path <- "ourdata/DataForStatistics_day0.csv"
# day2_path <- "ourdata/DataForStatistics_day2.csv"

# Loading data
metabs_day0 <- as.data.frame(read_csv(day0_path))
metabs_day2 <- as.data.frame(read_csv(day2_path))

# Changing first column into rownames. 
rownames(metabs_day0) <- metabs_day0$IsolabMidas
rownames(metabs_day2) <- metabs_day2$IsolabMidas

# Remove rownames column and Batch columns. 
metabs_day0 <- metabs_day0[,-c(1,12,13,14,15,16)]
metabs_day2 <- metabs_day2[,-c(1,11,12,13,14,15)] # this one only goes to 15 as CLL_01 is excluded.

# Currently, in day0 the column names are Control first, then CLL. In day2 it is visa versa. This is confusing. Let's fix this.
metabs_day0 <- metabs_day0[,c(6,7,8,9,10,1,2,3,4,5)]
# Now everything is CLL first, then Controls. 

# And remove CLL_01 of day0, as this one is excluded from RNA data too. 
metabs_day0 <- metabs_day0[,c(-1)]

# Also remove the day2 Control_03, as we do not have RNA data from this one.
metabs_day2 <- metabs_day2[,c(-7)]

# removing unnecessary variables / values
rm(day0_path)
rm(day2_path)



```

RNA
```{r}
# Path to data AMC PC
rawcounts_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\EC024_raw counts.csv"

# Path to data my laptop
# rawcounts_path <- "ourdata/EC024_raw counts.csv"


# Loading data, changing first column into rownames and removing this column.
RNA_rawcounts <- as.data.frame(read_csv(rawcounts_path))
rownames(RNA_rawcounts) <- RNA_rawcounts$...1
RNA_rawcounts <- RNA_rawcounts[,c(-1)]

# Split df into 2: day_0 and day_2.
# At the same time, the columns are being sorted: CLL first, in the same order as by metabolites
RNA_day0 <- RNA_rawcounts[,c(9,6,2,5,14,15,16,17,18)]
RNA_day2 <- RNA_rawcounts[,c(3,8,4,1,10,11,12,13)]

# Remove _T0 and _T48 parts of column names
# for (col in 1:ncol(RNA_day0)){
#   colnames(RNA_day0)[col] <- sub("_T0", "", colnames(RNA_day0)[col])
# }
# for (col in 1:ncol(RNA_day2)){
#   colnames(RNA_day2)[col] <- sub("_T48", "", colnames(RNA_day2)[col])
# }
# rm(col)

# Removing unnecessary variables
rm(rawcounts_path)
RNA <- RNA_rawcounts[,c(9,6,2,5,14,15,16,17,18,3,8,4,1,10,11,12,13)]

write.csv(RNA,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\all_RNA_raw.csv")

write.csv(RNA_day0,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day0_RNA_raw.csv")

write.csv(RNA_day2,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_raw.csv")


```

Metabolites again
```{r}
# Making a full metabolites dataframe

# Change metab df colnames to match the RNA df colnames
colnames(metabs_day0) <- colnames(RNA_day0)
colnames(metabs_day2) <- colnames(RNA_day2)

# Change colnames to include the day data, day 0 as _T0 and day 2 as _T48
# colnames(metabs_day0) <- paste0(colnames(metabs_day0), "_T0")
# colnames(metabs_day2) <- paste0(colnames(metabs_day2), "_T48")

# Add metabolite data back in, so they can be joined by...
metabs_day0$metabolites <- rownames(metabs_day0)
metabs_day2$metabolites <- rownames(metabs_day2)

# Create a file with all metab data
metabs <- as.data.frame(full_join(metabs_day0, metabs_day2, by = "metabolites"))
rownames(metabs) <- metabs$metabolites

# Rm col names metabolites
metabs_day0 <- subset(metabs_day0, select=-metabolites)
metabs_day2 <- subset(metabs_day2, select=-metabolites)
metabs <- subset(metabs, select=-metabolites)

# Change NA to 0, bc NA means undetected which means 0 for now? Might remove this later tho x
metabs[is.na(metabs)] <- 0


write.csv(metabs,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\all_metabs_raw.csv")

write.csv(metabs_day0,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\day0_metabs_raw.csv")

write.csv(metabs_day2,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\day2_metabs_raw.csv")

```


### Metabs - HMDB
Create file for metabolite matching

```{r}
# They have this one
# metabolite_matching <- as.data.frame(read_csv("Factor_COSMOS/support/metabolite_matching.csv"))

# We use this one - AMC Path
metabolite_matching <- as.data.frame(read_excel("L:/basic/divg/EXIM/ImmunoHematology/Cathy Magnée/Data/CD3_HDvsCLL/Metabolomic/Statistical Data_Incl_HMDB.xlsx"))
# We use this one - AMC Path
metabolite_matching <- as.data.frame(read_excel("ourdata/Statistical Data_Incl_HMDB.xlsx"))


metabolite_matching <- metabolite_matching[,c(1,5)]


```

For now, two methods for metabolite matching:
- Keep all
This means the measurements which can be applicable to multiple metabolites are being used for all possibilities.
- Keep none
This means the measurements which can be applicable to multiple metabolites are being used for none of the possibilities. 

```{r}
# Keep all of the duplicates (meaning the same measurements are being used for multiple metabolites).
metabolite_matching_all <- metabolite_matching

metabs_day0$Analyte.name <- rownames(metabs_day0)
metabs_day0_all <- merge(metabolite_matching_all, metabs_day0, by = "Analyte.name")
metabs_day2$Analyte.name <- rownames(metabs_day2)
metabs_day2_all <- merge(metabolite_matching_all, metabs_day2, by = "Analyte.name")


# Keep none of the duplicates (meaning the measurements of which it is unclear what metabolite is measured, are discarded).
metabolite_matching_none <- metabolite_matching[!(duplicated(metabolite_matching$Analyte.name) | duplicated(metabolite_matching$Analyte.name, fromLast = T)), ]
metabs_day0_none <- merge(metabolite_matching_none, metabs_day0, by = "Analyte.name")

metabs_day2_none <- merge(metabolite_matching_none, metabs_day2, by = "Analyte.name")

```

Note that this happens later in the 'classic' COSMOS pipeline.

### RNA - ENSEMBL and FPKM 

Tryout: Using Elena's FPKM data. Thank you Elena, very cool. 

```{r}
# Create FPKM df
RNA_FPKM <- RNA
RNA_FPKM$gene_id <- rownames(RNA_FPKM)

# Path to files containing FPKM values 
rsem_count.path <- "L:/basic/divg/EXIM/ImmunoHematology/Elena Camerini/Bioinformatics/EC024_CJ_RNAseq/Raw data from alignment/rsem_count"

# genes.results files
all_files <- list.files(path = rsem_count.path, pattern = "*.genes.results", full.names = TRUE)

for (col_name in colnames(RNA_rawcounts)) {
  
  matching_files <- grep(col_name, all_files, value=T)

  # If there is exactly one matching file, proceed
  if (length(matching_files) == 1) {
    file_path <- matching_files[1]
    
    # Load the file into a dataframe called temp
    temp <- read.csv(file_path, sep = "\t")
    
    RNA_FPKM <- RNA_FPKM %>%
      left_join(temp %>% select(gene_id, FPKM), by = "gene_id") %>%
      mutate(!!col_name := FPKM) %>%
      select(-FPKM)
    
  } else if (length(matching_files) > 1) {
    warning(paste("Multiple files found for column name:", col_name))
  } else {
    warning(paste("No file found for column name:", col_name))
  }
}

RNA_FPKM <- tibble::column_to_rownames(RNA_FPKM, "gene_id")

# Remove overbodige (redundant) variables
rm(col_name)
rm(temp)


RNA_FPKM_day0 <- RNA_FPKM[,c(1:9)]
RNA_FPKM_day2 <- RNA_FPKM[,c(10:18)]

write.csv(RNA_FPKM,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\all_RNA_FPKM.csv")

write.csv(RNA_FPKM_day0,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day0_RNA_FPKM.csv")

write.csv(RNA_FPKM_day2,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_FPKM.csv")

```

Using log2 to transform data into the input data typically used at the start of the COSMOS pipeline

```{r}
# Use function
RNA_log2_FPKM <- log2(RNA_FPKM)

# Change previous 0 (which turns to -Inf) back into 0
RNA_log2_FPKM[RNA_log2_FPKM==-Inf]=0

RNA_log2_FPKM_day0 <- RNA_log2_FPKM[,c(1:9)]
RNA_log2_FPKM_day2 <- RNA_log2_FPKM[,c(10:18)]

write.csv(RNA_log2_FPKM,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\all_RNA_log2_FPKM.csv")

write.csv(RNA_log2_FPKM_day0,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day0_RNA_log2_FPKM.csv")

write.csv(RNA_log2_FPKM_day2,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_log2_FPKM.csv")


```

ENSEMBL to Gene names

```{r}
# AMC PC
ENSEMBL_path <- "data/Transcriptomic/Ensembl_annotation.csv"

# laptop
# ENSEMBL_path <- "ourdata/Ensembl_annotation.csv"

# not needed:
# test <- readDNAStringSet("L:/basic/divg/EXIM/ImmunoHematology/Elena Camerini/Bioinformatics/EC024_CJ_RNAseq/Raw data from alignment/ref_genome/human_ensembl.n2g.idx.fa")
# subseq(test, start=c(1, 2, 3), end=c(3, 6, 5))
# seq_name = names(test)
# sequence = paste(test)
# df <- data.frame(seq_name, sequence)



# Loading data, changing first column into rownames and removing this column.
ENSEMBL_names <- as.data.frame(read_csv(ENSEMBL_path))
ENSEMBL_names <- ENSEMBL_names[,c(-1,-4)]

rm(ENSEMBL_path)

RNA_log2_FPKM$GENE_ID <- rownames(RNA_log2_FPKM)


test <- left_join(RNA_log2_FPKM, ENSEMBL_names, by = "GENE_ID")

non_unique_genes <- test$GENE[duplicated(test$GENE) | duplicated(test$GENE, fromLast = TRUE)]
unique(non_unique_genes)

# rownames(test) <- test$GENE

```

```{r}
##get biomart lists
list <- listEnsembl()
ensembl <- useEnsembl(biomart="genes", dataset = "hsapiens_gene_ensembl")
ensembl

filters <- listFilters(ensembl)
head(filters)

attributes <- listAttributes(ensembl)
head(attributes)

symbols <- getBM(attributes=c('hgnc_symbol', 'ensembl_gene_id', 'uniprot_gn_id'), 
                 filters = 'ensembl_gene_id', 
                 values = RNA_log2_FPKM$GENE_ID, mart = ensembl)

colnames(symbols) <- c("Gene_name", "GENE_ID", "Uniprot_ID") 

ENSEMBL_names2 <- symbols[,-3]

test <- left_join(RNA_log2_FPKM, ENSEMBL_names2, by = "GENE_ID")
test <- unique(test)



# rownames(test) <- test$Gene_name

non_unique_genes <- test$Gene_name[duplicated(test$Gene_name) | duplicated(test$Gene_name, fromLast = TRUE)]
unique(non_unique_genes)

```

