---
title: "1_Preprocess_Transcriptomics"
author: "Cathy Magnée"
date: "2024-04-19"
output: html_document
execute:
echo: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Preprocessing Transcriptomics / RNAseq

This R Markdown document is a annotated version of the R scripts 'Prepare_RNA.R' and 'RNA.R' written by the saezlab group. The scripts and data can be found on the saezlab / Factor_COSMOS public github page.

The 'Prepare_RNA.R' scripts can specifically be found at the following github page: Factor_COSMOS/scripts/RNA/Prepare_RNA.R at main · saezlab/Factor_COSMOS (github.com)

The 'RNA.R' scripts can be found on this github page:

Factor_COSMOS/scripts/RNA.R at no_MOFA · saezlab/Factor_COSMOS (github.com)

Source: Dugourd A, Kuppe C, Sciacovelli M, Gjerga E, Gabor A, Emdal KB, Vieira V, Bekker-Jensen DB, Kranz J, Bindels EMJ, Jesper V Olsen, Christian Frezza, Rafael Kramann, Julio Saez-Rodriguez et al (2021) Causal integration of multi-omics data with prior knowledge to generate mechanistic hypotheses. Mol Syst Biol 17: e9730

<https://doi.org/10.15252/msb.20209730>

<https://www.embopress.org/doi/full/10.15252/msb.20209730>

### Loading...

Loading libraries:

```{r, echo=False}
library(readxl)
library(readr)
library(pheatmap)

```

Loading data:

```{r}
# Fragments per kilobase per million reads log2(FPKM + 1) --> RNAseq, composite expression from https://discover.nci.nih.gov/cellminer/datasets.do (https://discover.nci.nih.gov/cellminer/loadDownload.do)

RNA_log2_FPKM <- as.data.frame(read_excel("data/RNA/RNA__RNA_seq_composite_expression.xls", skip = 10, sheet = 1))

RNA_log2_FPKM <- as.data.frame(read_csv("L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_FPKM.csv"))

rownames(RNA_log2_FPKM) <- RNA_log2_FPKM$...1
RNA_log2_FPKM <- RNA_log2_FPKM[,-1]
RNA_log2_FPKM <- log2(RNA_log2_FPKM)

test <- RNA_log2_FPKM
test[test==-Inf]=0

```

#### Reformatting

In the following chunk, the RNAseq measurements are extracted from the table (as columns 2:6 are entrez gene ids, chromosome, start, end and cytoband. Then, the gene name column is turned into rownames and removed. Lastly, the cell line names are made uniform.

```{r}
# Extract RNAseq measurements and align cell line names
RNA_log2_FPKM <- RNA_log2_FPKM[,-c(2:6)]
rownames(RNA_log2_FPKM) <- RNA_log2_FPKM[,1]
RNA_log2_FPKM <- RNA_log2_FPKM[,-1]
names(RNA_log2_FPKM) <- gsub(".*[:]","",names(RNA_log2_FPKM))
RNA_log2_FPKM <- RNA_log2_FPKM[,order(colnames(RNA_log2_FPKM))]
names(RNA_log2_FPKM) <- gsub("MDA-MB-231", "MDA-MB-231/ATCC", names(RNA_log2_FPKM))
names(RNA_log2_FPKM) <- gsub("RXF 393", "RXF-393", names(RNA_log2_FPKM))
```

Right now, this is Fragments per kilobase per million reads log2(FPKM + 1) --\> RNAseq, composite expression from <https://discover.nci.nih.gov/cellminer/datasets.do> ([https://discover.nci.nih.gov/cellminer/loadDownload.do)](https://discover.nci.nih.gov/cellminer/loadDownload.do))

### Preparing RNA data

### Plotting

```{r}
hist(as.numeric(unlist(RNA_log2_FPKM)), breaks = 1000) 
```

Removing outliers:

```{r}
RNA_log2_FPKM[RNA_log2_FPKM < 1] <- NA
hist(as.numeric(unlist(RNA_log2_FPKM)), breaks = 1000) 
```

Remove rows that have more than 45 NA's

```{r}
RNA_log2_FPKM <- RNA_log2_FPKM[rowSums(is.na(RNA_log2_FPKM)) < 45,]
# RNA_log2_FPKM <- RNA_log2_FPKM[rowSums(RNA_log2_FPKM == 0) < 45,]
hist(as.numeric(unlist(RNA_log2_FPKM)), breaks = 1000) 
```

## Only keep cell lines that are both in metabolomics and transcriptomics

We do not have to do this, as we have CLL patients and HDs, not cell lines.

```{r}

metabolomic_clean_vsn <- as.data.frame(read_csv("data/metabolomic/metabolomic_clean_vsn.csv"))

# Merge them based only on the columns (cell lines) that are in both

cell_lines <- intersect(names(metabolomic_clean_vsn[,-1]),names(RNA_log2_FPKM))

# Only keep omics that intersected?
metabolomic_clean_vsn <- metabolomic_clean_vsn[,c("metabolite",cell_lines)]


RNA_log2_FPKM <- RNA_log2_FPKM[,cell_lines]
RNA_log2_FPKM$gene <- row.names(RNA_log2_FPKM)

RNA_log2_FPKM <- RNA_log2_FPKM[,c(59,1:58)]

write_csv(RNA_log2_FPKM,file = "data/RNA/RNA_log2_FPKM_cleaned_common.csv")
write_csv(metabolomic_clean_vsn, file = "data/metabolomic/metabolomic_clean_vsn_common.csv")
```
