---
title: "1_Preprocess_Transcriptomics"
author: "Cathy Magn√©e"
date: "2024-04-19"
output: html_document
execute:
echo: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Preprocessing Transcriptomics / RNAseq

This R Markdown document is a translation of the Rscripts 'Prepare_RNA.R' and ; Analyze_RNA.R, from the cosmosR (specifically, the FACTOR_COSMOS/scripts/RNA) github page.

### Loading...

Loading libraries:

```{r, echo=False}
library(readr)

```


Loading data and reformatting.

```{r}
RNA_log2_FPKM <- as.data.frame(read_delim("data/RNA/RNA_log2_FPKM.csv",, delim = ",", escape_double = FALSE, trim_ws = TRUE))
# RNA_log2_FPKM <- RNA_log2_FPKM[,-c(2:6)]
names(RNA_log2_FPKM) <- gsub(".*[:]","",names(RNA_log2_FPKM))

RNA_log2_FPKM <- RNA_log2_FPKM[complete.cases(RNA_log2_FPKM),]

row.names(RNA_log2_FPKM) <- RNA_log2_FPKM$Genes
RNA_log2_FPKM <- RNA_log2_FPKM[,-1]

# RNA_log2_FPKM <- apply(RNA_log2_FPKM,2,function(x){
#   gsub(",",".",x)
# }) 

# RNA_log2_FPKM <- as.data.frame(RNA_log2_FPKM)

# Our data would be:
# RNA_log2_FPKM <- as.data.frame(read_excel(path="data/Transcriptomics/EC024_raw counts.xls", sheet = 1))

```

Right now, this is Fragments per kilobase per million reads log2(FPKM + 1) --\> RNAseq, composite expression from <https://discover.nci.nih.gov/cellminer/datasets.do> ([https://discover.nci.nih.gov/cellminer/loadDownload.do)](https://discover.nci.nih.gov/cellminer/loadDownload.do))


### Plotting 
```{r}
hist(as.numeric(unlist(RNA_log2_FPKM)), breaks = 1000) 
```


Removing outliers:
```{r}
RNA_log2_FPKM[RNA_log2_FPKM < 1] <- NA
hist(as.numeric(unlist(RNA_log2_FPKM)), breaks = 1000) 
```

Remove rows that have more than 45 NA's
```{r}
RNA_log2_FPKM <- RNA_log2_FPKM[rowSums(is.na(RNA_log2_FPKM)) < 45,]
# RNA_log2_FPKM <- RNA_log2_FPKM[rowSums(RNA_log2_FPKM == 0) < 45,]
hist(as.numeric(unlist(RNA_log2_FPKM)), breaks = 1000) 
```


## Only keep cell lines that are both in metabolomics and transcriptomics

We do not have to do this, as we have CLL patients and HDs, not cell lines. 
```{r}

metabolomic_clean_vsn <- as.data.frame(read_csv("data/metabolomic/metabolomic_clean_vsn.csv"))

# Merge them

cell_lines <- intersect(names(metabolomic_clean_vsn[,-1]),names(RNA_log2_FPKM))

# Only keep omics that intersected?
metabolomic_clean_vsn <- metabolomic_clean_vsn[,c("metabolite",cell_lines)]


RNA_log2_FPKM <- RNA_log2_FPKM[,cell_lines]
RNA_log2_FPKM$gene <- row.names(RNA_log2_FPKM)

RNA_log2_FPKM <- RNA_log2_FPKM[,c(59,1:58)]

write_csv(RNA_log2_FPKM,file = "data/RNA/RNA_log2_FPKM_cleaned_common.csv")
write_csv(metabolomic_clean_vsn, file = "data/metabolomic/metabolomic_clean_vsn_common.csv")
```

