---
title: "1 - Preparing omics data"
author: "Cathy Magnée"
date: "2024-06-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Preprocessing T cell omics data

This document describes the preprocessing of transcriptomic and metabolomic T cell data. It is based off the preprocessing documents in the FACTOR_Cosmos No_MOFA branch by Saezlab on github. This document is precedes the 2 - Preparing Cosmos inputs.Rmd document. 

This document is annotated to explain the steps and the choices that were made durnig this project. Other options will also be highlighted.

### Loading libraries
Installing packages (if you have not yet)

```{r}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("metaboliteIDmapping")
# BiocManager::install("vsn")
# BiocManager::install("biomaRt")
#
```

Loading libraries:
```{r}
library(readxl)
library(readr)
library(biomaRt)
library(dplyr)
library(reshape2)
library(pheatmap)
library(vsn)
library(metaboliteIDmapping)
```


### Transcriptomics

Step 1. Importing raw gene expression data (transcriptomic):
```{r}
############ Path to data AMC PC ############ 
rawcounts_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\EC024_raw counts.csv"

############ Path to data laptop ############ 
# rawcounts_path <- "../ourdata/EC024_raw counts.csv"

# Loading data, changing first column into rownames and removing this column.
RNA_rawcounts <- as.data.frame(read_csv(rawcounts_path))

# Make rownames, then remove them (only to put them back later)
rownames(RNA_rawcounts) <- RNA_rawcounts$...1
RNA_rawcounts <- RNA_rawcounts[,c(-1)]


# Order them so Day 0 and day 2 are separate. 
RNA <- RNA_rawcounts[,c(7,9,6,2,5,14,15,16,17,18,3,8,4,1,10,11,12,13)] # I am editing things
RNA$GENE_ID <- rownames(RNA)

# Remove redundant variables
rm(rawcounts_path)

```
Step 2. Importing the FPKM values
```{r}
### This is raw counts, we want the FPKM data:
# Prepare a FPKM df
RNA_FPKM <- RNA[,c(1:18)]
RNA_FPKM$gene_id <- rownames(RNA_FPKM)

# Path to files containing FPKM values 
############ AMC ############ 
rsem_count.path <- "L:/basic/divg/EXIM/ImmunoHematology/Elena Camerini/Bioinformatics/EC024_CJ_RNAseq/Raw data from alignment/rsem_count"

############ Laptop ############ 
# rsem_count.path <- "../ourdata/rsem_count"

# genes.results files
all_files <- list.files(path = rsem_count.path, pattern = "*.genes.results", full.names = TRUE)

for (col_name in colnames(RNA_rawcounts)) { 
  
  matching_files <- grep(col_name, all_files, value=T)

  # If there is exactly one matching file, proceed
  if (length(matching_files) == 1) {
    file_path <- matching_files[1]
    
    # Load the file into a dataframe called temp
    temp <- read.csv(file_path, sep = "\t")
    
    RNA_FPKM <- RNA_FPKM %>%
      left_join(temp %>% select(gene_id, FPKM), by = "gene_id") %>%
      mutate(!!col_name := FPKM) %>%
      select(-FPKM)
    
  } else if (length(matching_files) > 1) {
    warning(paste("Multiple files found for column name:", col_name))
  } else {
    warning(paste("No file found for column name:", col_name))
  }
}

# Remove redundant variables
rm(col_name)
rm(temp)
rm(all_files)
rm(matching_files)
rm(rsem_count.path)
rm(file_path)

```

Step 3. Transforming ENSEMBL Gene IDs to gene names using biomaRt
```{r}

# Prepare Ensembl names
list <- listEnsembl()
ensembl <- useEnsembl(biomart="genes", dataset = "hsapiens_gene_ensembl")

filters <- listFilters(ensembl)
attributes <- listAttributes(ensembl)

symbols <- getBM(attributes=c('hgnc_symbol', 
                              'ensembl_gene_id', 
                              'uniprot_gn_id'), 
                 filters = 'ensembl_gene_id', 
                 values = RNA$GENE_ID, mart = ensembl)

colnames(symbols) <- c("Gene_name", "gene_id", "Uniprot_ID") 

ENSEMBL_names2 <- symbols[,-3]
ENSEMBL_names2 <- unique(ENSEMBL_names2)

# Add Ensembl Gene names
RNA_FPKM <- left_join(RNA_FPKM, ENSEMBL_names2, by = "gene_id")

# Filter out the non-unique gene names
non_unique_genes <- RNA_FPKM$Gene_name[duplicated(RNA_FPKM$Gene_name) | duplicated(RNA_FPKM$Gene_name, fromLast = TRUE)]
unique(non_unique_genes)
RNA_FPKM <- RNA_FPKM[!RNA_FPKM$Gene_name %in% RNA_FPKM$Gene_name[duplicated(RNA_FPKM$Gene_name)], ]

# Check if there are no more non unique genes. 
non_unique_genes <- RNA_FPKM$Gene_name[duplicated(RNA_FPKM$Gene_name) | duplicated(RNA_FPKM$Gene_name, fromLast = TRUE)] # Should be empty now

# Remove redundant variables
rm(non_unique_genes)
rm(ensembl)
rm(filters)
rm(list)

# Remove column containing the gene names
rownames(RNA_FPKM) <- RNA_FPKM$Gene_name
RNA_FPKM <- RNA_FPKM[,c(1:18)]
# RNA_FPKM <- tibble::column_to_rownames(RNA_FPKM, "Gene_name")

# Create day 0 and day 2 dfs
RNA_FPKM_day0 <- RNA_FPKM[,c(1:10)]
RNA_FPKM_day2 <- RNA_FPKM[,c(11:18)]

# Save the data
############ AMC ############ 
write.csv(RNA_FPKM,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\all_RNA_FPKM.csv")

write.csv(RNA_FPKM_day0,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day0_RNA_FPKM.csv")

write.csv(RNA_FPKM_day2,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_FPKM.csv")

############ Laptop ############ 
# write.csv(RNA_FPKM,file = "../ourdata/transcriptomics/all_RNA_FPKM.csv")
# 
# write.csv(RNA_FPKM_day0,file = "../ourdata/transcriptomics/day0_RNA_FPKM.csv")
# 
# write.csv(RNA_FPKM_day2,file = "../ourdata/transcriptomics/day2_RNA_FPKM.csv")

```

Step 4. Log2 transformation 
```{r}
# Use function
RNA_log2_FPKM <- log2(RNA_FPKM)

# Change previous 0 (which turns to -Inf) back into 0
RNA_log2_FPKM[RNA_log2_FPKM==-Inf]=0

RNA_log2_FPKM_day0 <- RNA_log2_FPKM[,c(1:10)]
RNA_log2_FPKM_day2 <- RNA_log2_FPKM[,c(11:18)]


### Saving these files again, separately 

############ AMC ############ 
write.csv(RNA_log2_FPKM,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\all_RNA_log2_FPKM.csv")

write.csv(RNA_log2_FPKM_day0,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day0_RNA_log2_FPKM.csv")

write.csv(RNA_log2_FPKM_day2,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_log2_FPKM.csv")

############ Laptop ############ 
# write.csv(RNA_FPKM,file = "../ourdata/transcriptomics/all_RNA_log2_FPKM.csv")
# 
# write.csv(RNA_FPKM_day0,file = "../ourdata/transcriptomics/day0_RNA_log2_FPKM.csv")
# 
# write.csv(RNA_FPKM_day2,file = "../ourdata/transcriptomics/day2_RNA_log2_FPKM.csv")
```
Log2 transformed data is the typical input of COSMOS. 

### Metabolomics 

Step 1. Importing metabolite data and editing columns 
Also excluding some columns
```{r}

# Path do data: AMC PC
day0_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\DataForStatistics_day0.csv"
day2_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\DataForStatistics_day2.csv"

# Path to data laptop
# day0_path <- "../ourdata/DataForStatistics_day0.csv"
# day2_path <- "../ourdata/DataForStatistics_day2.csv"

# Loading data
metabs_day0 <- as.data.frame(read_csv(day0_path))
metabs_day2 <- as.data.frame(read_csv(day2_path))

# Changing first column into rownames. 
rownames(metabs_day0) <- metabs_day0$IsolabMidas
rownames(metabs_day2) <- metabs_day2$IsolabMidas

# Remove 'rownames column' and 'Batch' columns. 
metabs_day0 <- metabs_day0[,-c(1,12,13,14,15,16)]
metabs_day2 <- metabs_day2[,-c(1,11,12,13,14,15)] 

# Fix day0 order of columns (CLL first, then control)
metabs_day0 <- metabs_day0[,c(6,7,8,9,10,1,2,3,4,5)]

# Remove day0 CLL_01, as there is no gene expression data available
metabs_day0 <- metabs_day0[,c(-1)]

# Remove day2 Control_03, as there is no gene expression data available
metabs_day2 <- metabs_day2[,c(-7)]

# removing unnecessary variables / values
rm(day0_path)
rm(day2_path)

# Save these files
#### AMC
write.csv(metabs_day0,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\day0_metabs_raw.csv")

write.csv(metabs_day2,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Metabolomic\\day2_metabs_raw.csv")


#### Laptop
write.csv(metabs_day0,file = "../ourdata/metabolomic/day0_metabs_raw.csv")

write.csv(metabs_day2,file = "../ourdata/metabolomic/day2_metabs_raw.csv")

```

Step 2. Editing metabolite names to HMDB ID's
- Importing Table for matching metabolite names to HMDB ID's
```{r}
# Metabolite name to HMDB table
# AMC Path
metabolite_matching <- as.data.frame(read_excel("L:/basic/divg/EXIM/ImmunoHematology/Cathy Magnée/Data/CD3_HDvsCLL/Metabolomic/Statistical Data_Incl_HMDB.xlsx"))

# Laptop Path
metabolite_matching <- as.data.frame(read_excel("ourdata/Statistical Data_Incl_HMDB.xlsx"))

metabolite_matching <- metabolite_matching[,c(1,5)]

```

There will be duplicates, just like in ENSEMBL. And just like in ENSEMBL, we will remove all duplicates:

```{r}
# Keep none of the duplicates (
# This means all measurements are discarded when it is unclear what metabolite is measured 

metabolite_matching_none <- metabolite_matching[!(duplicated(metabolite_matching$Analyte.name) | duplicated(metabolite_matching$Analyte.name, fromLast = T)), ]


metabs_day0 <- merge(metabolite_matching_none, metabs_day0, by = "Analyte.name")

metabs_day2 <- merge(metabolite_matching_none, metabs_day2, by = "Analyte.name")
```

Note that this happens later in the 'classic' FACTOR_Cosmos pipeline.

Step 3. Saving data
```{r}

write.csv(metabs_day0,file = "../ourdata/metabolomic/day0_metabs_hmdb.csv")

write.csv(metabs_day2,file = "../ourdata/metabolomic/day2_metabs_hmdb.csv")
```


