---
title: "2 - Preprocessing gene expression data"
author: "Cathy Magnée, Saezlab"
date: "2024-06-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Preprocessing T cell transcriptomics data

This document describes the preprocessing of transcriptomic and metabolomic T cell data. It is based off the preprocessing documents in the FACTOR_Cosmos No_MOFA branch by Saezlab on github. This document is preceded by the 1 - Preparing omics data.Rmd document and precedes the 3 - PReprocessing metabolite R markdown document. 

This document is annotated to explain the steps and the choices that were made durnig this project. Other options will be highlighted.

### Loading libraries
Installing packages (if you have not yet)

```{r}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("vsn")
# BiocManager::install("biomaRt")

```

Loading libraries:
```{r}
library(readxl)
library(readr)
library(pheatmap)
library(tibble)
```

Loading data
```{r}

# Defining paths
############ AMC ############  
day0_RNA_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day0_RNA_log2_FPKM.csv"

day2_RNA_path <- "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_log2_FPKM.csv"

############ Laptop ############ 
# day0_RNA_path <- "../ourdata/transcriptomic/day0_RNA_log2_FPKM.csv"
# day2_RNA_path <- "../ourdata/transcriptomic/day2_RNA_log2_FPKM.csv"

# day0_RNA_path <- "../ourdata/transcriptomic/day0_RNA_raw.csv"
# day2_RNA_path <- "../ourdata/transcriptomic/day2_RNA_raw.csv"

# Loading data
day0_RNA_log2_FPKM <- as.data.frame(read_csv(day0_RNA_path))

day2_RNA_log2_FPKM <- as.data.frame(read_csv(day2_RNA_path))

# Change first column to rownames
day0_RNA_log2_FPKM <- tibble::column_to_rownames(day0_RNA_log2_FPKM, var = "...1")
day2_RNA_log2_FPKM <- tibble::column_to_rownames(day2_RNA_log2_FPKM, var = "...1")

# Remove redundant variables
rm(day0_RNA_path)
rm(day2_RNA_path)

```

### Create function to clean data

```{r}
# Cleaning function
clean_gene_expression_data <- function(RNA_log2_FPKM) {
  # Everything less than 1 should be NA
  RNA_log2_FPKM[RNA_log2_FPKM < 1] <- NA 
  hist(as.numeric(unlist(RNA_log2_FPKM)), breaks = 1000)
  # All genes that have no expression in > 20% of samples should be removed
  RNA_log2_FPKM[rowSums(is.na(RNA_log2_FPKM)) < 0.8*ncol(RNA_log2_FPKM),]
}

# Scaling function
scale_gene_expression_data <- function(RNA_log2_FPKM_cleaned) {
  # Grab Standard Deviation
  SDs <- apply(RNA_log2_FPKM_cleaned,1,function(x){sd(x,na.rm = T)})
  # Grab mean
  means <- rowMeans(RNA_log2_FPKM_cleaned, na.rm = T)
  # Z-normalization 
  RNA_log2_FPKM_cleaned <- (RNA_log2_FPKM_cleaned - means) / SDs
}
```

```{r}
# hist(as.numeric(unlist(day0_RNA_log2_FPKM)), breaks = 1000)
# day0_RNA_log2_FPKM_cleaned <- clean_gene_expression_data(day0_RNA_log2_FPKM)
# hist(as.numeric(unlist(day0_RNA_log2_FPKM_cleaned)), breaks = 1000)
# 
# day0_RNA_log2_FPKM_cleaned <- scale_gene_expression_data(day0_RNA_log2_FPKM_cleaned)
# hist(as.numeric(unlist(day0_RNA_log2_FPKM_cleaned[,c(1:8)])), breaks = 1000)
# hist(as.numeric(unlist(day0_RNA_log2_FPKM_cleaned[,c(1:4)])), breaks = 1000)
# hist(as.numeric(unlist(day0_RNA_log2_FPKM_cleaned[,c(5:8)])), breaks = 1000)

```


### Plotting data

Day 0
```{r}
hist(as.numeric(unlist(day0_RNA_log2_FPKM)), breaks = 1000)
hist(as.numeric(unlist(day2_RNA_log2_FPKM)), breaks = 1000)

```

### Filtering Data
Many <1, so let's remove those and plot again:
```{r}

day0_RNA_log2_FPKM[day0_RNA_log2_FPKM < 1] <- NA
hist(as.numeric(unlist(day0_RNA_log2_FPKM)), breaks = 1000) 

day2_RNA_log2_FPKM[day2_RNA_log2_FPKM < 1] <- NA
hist(as.numeric(unlist(day2_RNA_log2_FPKM)), breaks = 1000) 
```

Let's remove all rows (genes) where >20% is NA
For day0, where there are 10 samples, this means to keep all samples where <8 out of 10 are NA.
For day2, this means to keep samples where less than 6.4 out of 8 are NA

```{r}
day0_RNA_log2_FPKM <- day0_RNA_log2_FPKM[rowSums(is.na(day0_RNA_log2_FPKM)) < 0.8*ncol(day0_RNA_log2_FPKM),]

hist(as.numeric(unlist(day0_RNA_log2_FPKM)), breaks = 1000) 

day2_RNA_log2_FPKM <- day2_RNA_log2_FPKM[rowSums(is.na(day2_RNA_log2_FPKM)) < 0.8*ncol(day2_RNA_log2_FPKM),]
hist(as.numeric(unlist(day2_RNA_log2_FPKM)), breaks = 1000) 

# hist(as.numeric(unlist(day2_RNA_log2_FPKM[,c(1:4)])), breaks = 1000)
# hist(as.numeric(unlist(day2_RNA_log2_FPKM[,c(5:8)])), breaks = 1000)

# day2_RNA_log2_FPKM <- scale_gene_expression_data(day2_RNA_log2_FPKM)
# 
# hist(as.numeric(unlist(day2_RNA_log2_FPKM)), breaks = 100)
# 
# day0_RNA_log2_FPKM <- scale_gene_expression_data(day0_RNA_log2_FPKM)
# 
# hist(as.numeric(unlist(day0_RNA_log2_FPKM[,c(1:4)])), breaks = 1000)
# 
# hist(as.numeric(unlist(day0_RNA_log2_FPKM[,c(5:8)])), breaks = 1000)


```

### Saving data
AMC
```{r}
write.csv(day0_RNA_log2_FPKM,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day0_RNA_log2_FPKM_cleaned.csv")
write.csv(day2_RNA_log2_FPKM,file = "L:\\basic\\divg\\EXIM\\ImmunoHematology\\Cathy Magnée\\Data\\CD3_HDvsCLL\\Transcriptomic\\day2_RNA_log2_FPKM_cleaned.csv")
```

Laptop
```{r}

# write.csv(day0_RNA_log2_FPKM,file = "../ourdata/transcriptomic/day0_RNA_log2_FPKM_cleaned.csv")
# write.csv(day2_RNA_log2_FPKM,file = "../ourdata/transcriptomic/day2_RNA_log2_FPKM_cleaned.csv")
```

